include/envoy/config/watch_map.h

```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <set>
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/config/subscription.h"
+
+namespace Envoy {
+namespace Config {
+
+struct Watch;
+using WatchPtr = std::unique_ptr<Watch>;
+
+struct AddedRemoved {
+  AddedRemoved(std::set<std::string> added, std::set<std::string> removed)
+      : added_(added), removed_(removed) {}
+  std::set<std::string> added_;
+  std::set<std::string> removed_;
+};
+
+class WatchMap {
+public:
+  virtual ~WatchMap() = default;
+
+  // Adds 'callbacks' to the WatchMap, with no resource names being watched.
+  // (Use updateWatchInterest() to add some names).
+  // Returns the newly added watch, to be used for updateWatchInterest. Destroy to remove from map.
+  virtual WatchPtr addWatch(SubscriptionCallbacks& callbacks) PURE;
+
+  // Updates the set of resource names that the given watch should watch.
+  // Returns any resource name additions/removals that are unique across all watches. That is:
+  // 1) if 'resources' contains X and no other watch cares about X, X will be in added_.
+  // 2) if 'resources' does not contain Y, and this watch was the only one that cared about Y,
+  //    Y will be in removed_.
+  virtual AddedRemoved updateWatchInterest(Watch* watch,
```

Comment:
See my comment below; I think if we go to RAII there, then it would make sense to move the `updateWatchInterest` method to the `Watch` itself.

Reply:
I can see how updateWatchInterest() itself would look cleaner with `Watch* watch` becoming an implicit `this`. However, I'm not sure if it's a net cleanliness gain: it would mean growing Watch from a simple struct to another full-fledged class that the reader has to understand. I would say, wait until this code is submitted, and look at this change as a standalone PR, and then see if it feels worth it.

Reply:
Since `Watch` encapsulates (at an interface level) the concept of a watch, it seems conceptually clean to me to have its mutation operations, e.g. `update` expressed directly on it. But, I freely admit that I might be wrong and would defer to another reviewer on this, e.g. @mattklein123.

Reply:
I'm still in favor of making this a direct member method on `Watch` FWIW.

Reply:
Done.